tests/client_api.cpp

```diff
@@ -1575,3 +1575,79 @@ TEST(Groups, Profiles)
 
         alice->close();
 }
+
+TEST(ClientAPI, PublicRooms)  
+{
+        // Setup : Create a new (public) room with some settings, add a user to the room
+        auto alice = std::make_shared<Client>("localhost");
+        auto bob   = std::make_shared<Client>("localhost");
+
+        alice->login("alice", "secret", [alice](const mtx::responses::Login &, RequestErr err) {
+                check_error(err);
+        });
+
+        bob->login("bob", "secret", [bob](const mtx::responses::Login &, RequestErr err) {
+                check_error(err);
+        });
+
+        while (alice->access_token().empty() || bob->access_token().empty())
+                sleep();
+
+        mtx::requests::CreateRoom req;
+        req.name = "Public Room";
+        req.topic = "Test";
+        req.visibility = Visibility::Public;
+        req.invite = {"@bob:localhost"};
+
+        alice->create_room(
+                req, [alice, bob](const mtx::responses::CreateRoom &res, RequestErr err){
+                check_error(err);
+                auto room_id = res.room_id.to_string();
+                bob->join_room(room_id, [](const mtx::responses::RoomId &, RequestErr err) {
+                        check_error(err);
+                });
+
+                // TEST 1: endpoints to set and get the visibility of the room we just created
+                json body = {{"visibility", "public"}};
+                alice->put_room_visibility(room_id, body, [alice](RequestErr err){
+                        check_error(err);
+                });
+
+                alice->get_room_visibility
+                (room_id, [alice](const mtx::responses::RoomVisibility &res, RequestErr err){
+                        check_error(err);
+                        EXPECT_EQ(visibilityToString(res.visibility), "public");
+                });
+
+                alice->get_room_visibility
+                ("", [alice](const mtx::responses::RoomVisibility &, RequestErr err) {
+                        ASSERT_TRUE(err);
+                        EXPECT_EQ(mtx::errors::to_string(err->matrix_error.errcode),
+                        "M_NOT_FOUND");
+                });
+
+                // TEST 2: endpoints to add and list the public rooms on the server 
+                mtx::requests::PublicRooms room_req;
+                room_req.limit = 1;
+                room_req.include_all_networks = true;
+                json j = room_req;
+                std::cout << j.dump(4) << "\n";
+
+                alice->post_public_rooms
+                (room_req, [alice](const mtx::responses::PublicRooms &, RequestErr err) {
+                        check_error(err);
+                });
```

Comment:
This will fail, because you execute the requests at the same time. You need to execute the next request in the callback of the one before it, otherwise all the gets will fail.